### PR TITLE
HER Wrappers

### PR DESCRIPTION
**Pull request by hades-rp2010**
Wrt #171 
Have added a `HERTrainer`, `HERGoalEnvWrapper`, and a `HERWrapper` for the replay buffer.
Some changes in the locations of the tests might be needed.. Wasnt too sure of where to put them

### CODE REVIEW

File: genrl/trainers/her_trainer.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
+from typing import Type, Union
+
+from genrl.trainers import OffPolicyTrainer
+
+
+class HERTrainer(OffPolicyTrainer):
```
Comment (Sharad24) [inline genrl/trainers/her_trainer.py:6]: Might need to refactor this -- arent there functions here which are exact duplicates of the ones in `OffPolicyTrainer`?

File: tests/test_deep/test_agents/test_her.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+from genrl.agents import DQN
+from genrl.core import ReplayBuffer
+from genrl.core.buffers import HERWrapper
+from genrl.environments.her_wrapper import HERGoalEnvWrapper
+from genrl.trainers import OffPolicyTrainer
+from genrl.trainers.her_trainer import HERTrainer
+
+
+class BitFlippingEnv(GoalEnv):
```
Comment (Sharad24) [inline tests/test_deep/test_agents/test_her.py:16]: Should this be in the tests?

Comment (Sharad24) [reply tests/test_deep/test_agents/test_her.py:16]: Maybe we could move this somewhere in the `environments` module.

**Review by Sharad24 — COMMENTED** [review]
Awesome work! @hades-rp2010 

Few questions:
1. Have you tried training? 
2. Does the wrapper work with all off policy algorithms?
3. Have you checked out https://github.com/eleurent/highway-env? Highway is a standard goal based env.

**Comment by hades-rp2010** [comment]
> Awesome work! @hades-rp2010 
> 
> Few questions:
> 1. Have you tried training? 
> 2. Does the wrapper work with all off policy algorithms?
> 3. Have you checked out https://github.com/eleurent/highway-env? Highway is a standard goal based env.

1. Still working on this, should mostly be over soon.
2. Yeah, my latest commit adds tests for all the off policy agents
3. Added tests for highway env too

**Comment by Sharad24** [comment]
Add `highway_env` to the `pip install ...` statements in `.github/workflows/tests.yml ` and `.github/workflows/codecov.yml`, should work post that.

**Comment by sampreet-arthi** [comment]
Are you done here? @hades-rp2010 If you can resolve the merge conflicts and maybe the codeclimate issues then we can merge this.

**Comment by Sharad24** [comment]
I think the algorithms weren’t training here. Please ensure you get a reasonable reward :)
